File: golang/cmd/day4/main.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"strconv"
	"strings"

	"github.com/ventsislav-georgiev/advent-of-code-22/golang/pkg/aoc"
)

func main() {
	aoc.Exec(task1, task2)
}

func task1(in io.ReadCloser) {
	scanner := bufio.NewScanner(in)

	result := 0
	for scanner.Scan() {
		pairs := strings.Split(scanner.Text(), ",")
		a := strings.Split(pairs[0], "-")
		b := strings.Split(pairs[1], "-")
		a_x, _ := strconv.Atoi(a[0])
		a_y, _ := strconv.Atoi(a[1])
		b_x, _ := strconv.Atoi(b[0])
		b_y, _ := strconv.Atoi(b[1])
		if (a_x <= b_x && a_y >= b_y) || (b_x <= a_x && b_y >= a_y) {
			result++
		}
	}

	fmt.Println(result)
}

func task2(in io.ReadCloser) {
	scanner := bufio.NewScanner(in)

	result := 0
	for scanner.Scan() {
		pairs := strings.Split(scanner.Text(), ",")
		a := strings.Split(pairs[0], "-")
		b := strings.Split(pairs[1], "-")
		a_x, _ := strconv.Atoi(a[0])
		a_y, _ := strconv.Atoi(a[1])
		b_x, _ := strconv.Atoi(b[0])
		b_y, _ := strconv.Atoi(b[1])
		if (a_x <= b_x && a_y >= b_x) || (b_x <= a_x && b_y >= a_x) {
			result++
		}
	}

	fmt.Println(result)
}

File: rust/day4/src/main.rs
use aoc::exec;

fn main() {
    exec(&[task1, task2]);
}

fn task1(input: String) {
    let mut result = 0;
    for line in input.lines() {
        let pairs: Vec<&str> = line.split(",").collect();
        let a: Vec<&str> = pairs[0].split("-").collect();
        let b: Vec<&str> = pairs[1].split("-").collect();
        let a_x = a[0].parse::<i32>().unwrap();
        let a_y = a[1].parse::<i32>().unwrap();
        let b_x = b[0].parse::<i32>().unwrap();
        let b_y = b[1].parse::<i32>().unwrap();
        if (a_x <= b_x && a_y >= b_y) || (b_x <= a_x && b_y >= a_y) {
            result += 1;
        }
    }

    println!("{}", result);
}

fn task2(input: String) {
    let mut result = 0;
    for line in input.lines() {
        let pairs: Vec<&str> = line.split(",").collect();
        let a: Vec<&str> = pairs[0].split("-").collect();
        let b: Vec<&str> = pairs[1].split("-").collect();
        let a_x = a[0].parse::<i32>().unwrap();
        let a_y = a[1].parse::<i32>().unwrap();
        let b_x = b[0].parse::<i32>().unwrap();
        let b_y = b[1].parse::<i32>().unwrap();
        if (a_x <= b_x && a_y >= b_x) || (b_x <= a_x && b_y >= a_x) {
            result += 1;
        }
    }

    println!("{}", result);
}
